src/compiler/emitter.ts

```diff
@@ -168,155 +167,60 @@ module ts {
                 return !resolver.hasGlobalName(name) &&
                     !hasProperty(currentSourceFile.identifiers, name) &&
                     (!generatedNameSet || !hasProperty(generatedNameSet, name))
             }
 
-            // in cases like
-            // for (var x of []) {
-            //     _i;
-            // }
-            // we should be able to detect if let _i was shadowed by some temp variable that was allocated in scope
-            function nameConflictsWithSomeTempVariable(name: string): boolean {
-                // temp variable names always start with '_'
-                if (name.length < 2 || name.charCodeAt(0) !== CharacterCodes._) {
-                    return false;
-                }
-
-                if (name === "_i") {
-                    return !!(predefinedTempsInUse & TempVariableKind._i);
-                }
-
-                if (name === "_n") {
-                    return !!(predefinedTempsInUse & TempVariableKind._n);
-                }
-
-                if (name.length === 2 && name.charCodeAt(1) >= CharacterCodes.a && name.charCodeAt(1) <= CharacterCodes.z) {
-                    // handles _a .. _z
-                    let n = name.charCodeAt(1) - CharacterCodes.a;
-                    return n < tempCount;
-                }
-                else {
-                    // handles _1, _2...
-                    let n = +name.substring(1);
-                    return !isNaN(n) && n >= 0 && n < (tempCount - 26);
-                }
-            }
-
-            // This function generates a name using the following pattern:
-            // _a .. _h, _j ... _z, _0, _1, ...
-            // It is guaranteed that generated name will not shadow any existing user-defined names,
-            // however it can hide another name generated by this function higher in the scope.
-            // NOTE: names generated by 'makeTempVariableName' and 'makeUniqueName' will never conflict.
-            // see comment for 'makeTempVariableName' for more information.
-            function makeTempVariableName(location: Node, tempVariableKind: TempVariableKind): string {
-                let tempName: string;
-                if (tempVariableKind !== TempVariableKind.auto && !(predefinedTempsInUse & tempVariableKind)) {
-                    tempName = tempVariableKind === TempVariableKind._i ? "_i" : "_n";
-                    if (!resolver.resolvesToSomeValue(location, tempName)) {
-                        predefinedTempsInUse |= tempVariableKind;
-                        return tempName;
+            // Return the next available name in the pattern _a ... _z, _0, _1, ...
+            // TempFlags._i or TempFlags._n may be used to express a preference for that dedicated name.
+            // Note that names generated by makeTempVariableName and makeUniqueName will never conflict.
+            function makeTempVariableName(flags: TempFlags): string {
+                if (flags && !(tempFlags & flags)) {

```

Comment:
Oh it's because `flags` will be falsy if it is Auto. In that case, can you do `if (flags !== TempFlags.Auto && !(tempFlags & flags))`.

Could you also rename flags to `desiredFlags` or something to indicate that it's the flag of the name you are trying to create?